### PR TITLE
blog: libtailscale internals

### PR DESCRIPTION
**Pull request by crawshaw**
This should cross-link back to a tailscale.com announce post, but I guess one has to come first, so I guess I'll put this one up first and then edit it later to link back.

### CODE REVIEW

File: data/blog/libtailscale.mdx [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+But it is the nature of an unmanaged heap like C has that programmers
+have to be responsible for freeing memory.
+
+## Future work
+
+There are two next steps for libtailscale.
+The first is writing libraries in other languages on top of their
+foreign function interface libraries.
+At some point, in some way, every language talks to C.
+It should be possible to put a `tailscale.jar` file into your Java
```
Comment (shayne) [inline data/blog/libtailscale.mdx:181]: Should we mention [tsconnect](https://github.com/tailscale/tailscale/tree/main/cmd/tsconnect) here somewhere?

Comment (crawshaw) [reply data/blog/libtailscale.mdx:181]: Oh yes! Look at that, an example where you don't go through C. :) Will add this afternoon.

Comment (crawshaw) [reply data/blog/libtailscale.mdx:183]: Linked to Mihai's blog post about the ssh console as it I think is our best description of how it works.

File: data/blog/libtailscale.mdx [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+authors: ['crawshaw']
+summary: 'libtailscale: a Tailscale C library'
+---
+
+We just released an experimental libtailscale, the Tailscale C library.
+It is built on the Go [tsnet](https://pkg.go.dev/tailscale.com/tsnet)
+package which uses a userland network stack to implement TCP/IP
+directly inside your process.
+
+```
```
Comment (shayne) [inline data/blog/libtailscale.mdx:17]: For syntax highlighting: \```c

File: data/blog/libtailscale.mdx [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+[build modes](https://pkg.go.dev/cmd/go#hdr-Build_modes).
+
+## Go Build Modes
+
+The two key build modes are `c-archive` and `c-shared`.
+The first creates an object archive that can be linked by a standard
+C toolchain into a binary.
+The second creates a shared object (on linux, the .so file) that can
+be dynamically linked into a binary.
+They are surprisingly easily to use:
```
Comment (shayne) [inline data/blog/libtailscale.mdx:56]: s/easily/easy

File: data/blog/libtailscale.mdx [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+It’s not widely used, but the standard Go toolchain includes all the
+machinery necessary for producing a C library in the form of
+[build modes](https://pkg.go.dev/cmd/go#hdr-Build_modes).
+
+## Go Build Modes
+
+The two key build modes are `c-archive` and `c-shared`.
+The first creates an object archive that can be linked by a standard
+C toolchain into a binary.
+The second creates a shared object (on linux, the .so file) that can
```
Comment (shayne) [inline data/blog/libtailscale.mdx:54]: Linux

File: data/blog/libtailscale.mdx [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+```
+// servers tracks all the allocated *tsnet.Server objects.
+var servers struct {
+        mu   sync.Mutex
+        next C.int
+        m    map[C.int]*tsnet.Server
+}
+```
+
+An Go function `//export`ed to C can then allocate a new `*tsnet.Server`,
```
Comment (shayne) [inline data/blog/libtailscale.mdx:165]: While I love it, maybe "A Go"

File: data/blog/libtailscale.mdx [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+This lets C refer to a Go pointer, but for the pointer itself to
+remain entirely managed by the Go runtime.
+
+We use process-global maps like this for the `tailscale`,
+`tailscale_listener`, and `tailscale_conn` handles.
+
+This puts C completely in charge of freeing memory.
+If the C program does not call the appropriate close function, the Go
+memory will live forever in the map and not be collected.
+But it is the nature of an unmanaged heap like C has that programmers
```
Comment (shayne) [inline data/blog/libtailscale.mdx:179]: I struggled parsing this, suggest adding commas:

...heap, like C has, that ...

File: data/blog/libtailscale.mdx [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+There are two next steps for libtailscale.
+The first is writing libraries in other languages on top of their
+foreign function interface libraries.
+At some point, in some way, _almost_ every language talks to C.
+(A great counter-example is the
+[WASM Tailscale client](https://tailscale.com/blog/ssh-console/)).
+It should be possible to put a `tailscale.jar` file into your Java
+project and start dialing and listening on your tailnet.
+
+The second step is exposing more functionality of the tailscale node
```
Comment (shayne) [inline data/blog/libtailscale.mdx:193]: Tailscale

File: data/blog/libtailscale.mdx [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+to both C and other languages built on top of libtailscale.
+To do that we need to implement LocalAPI.
+There are a few interesting lessons in using cross-language bindings
+to be found there too, which should make for a nice followup blog post
```
Comment (shayne) [inline data/blog/libtailscale.mdx:197]: follow-up

File: data/blog/libtailscale.mdx [hunk 9]
```diff
@@ -0,0 +1,198 @@
+---
+title: "libtailscale: how it's made"
+date: '2023-03-02'
+tags:
+  - 'tsnet'
+  - 'libtailscale'
+  - 'oss'
+authors: ['crawshaw']
+summary: 'libtailscale: a Tailscale C library'
```
Comment (shayne) [inline data/blog/libtailscale.mdx:9]: Something longer. We're currently using these as the excerpts on the homepage.

Suggest:
"libtailscale: a Tailscale C library. Using the Go toolchain to build a c-archive tsnet package that uses a userland network stack to implement TCP/IP inside a process."

File: data/blog/libtailscale.mdx (hunk 5 above)
Comment (WalterHub) [inline data/blog/libtailscale.mdx:165]:
```suggestion
A Go function `//export`-ed to C can then allocate a new `*tsnet.Server`,
```

Comment (WalterHub) [reply data/blog/libtailscale.mdx:165]: Check to see whether the rendered `//export`-ed phrase introduces an extra space (between the code portion and the plain text portion).

File: data/blog/libtailscale.mdx [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+be dynamically linked into a binary.
+They are surprisingly easily to use:
+
+    go build -buildmode=c-archive pkg
+
+This produces an object archive, and a generated header for the C
+symbols Go exports.
+
+As part of compilation, cgo can “export” your Go functions to C using
+a `//export` directive.
```
Comment (WalterHub) [inline data/blog/libtailscale.mdx:64]: Use "an `//export`" here (instead of "a `//export`")?

File: data/blog/libtailscale.mdx [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+The second creates a shared object (on linux, the .so file) that can
+be dynamically linked into a binary.
+They are surprisingly easily to use:
+
+    go build -buildmode=c-archive pkg
+
+This produces an object archive, and a generated header for the C
+symbols Go exports.
+
+As part of compilation, cgo can “export” your Go functions to C using
```
Comment (WalterHub) [inline data/blog/libtailscale.mdx:63]: Curious to know if the use of curly quotes in “export” render properly.  Might be fine, just something to check.

File: data/blog/libtailscale.mdx (hunk 2 above)
Comment (shayne) [reply data/blog/libtailscale.mdx:17]: Offer a description of the sample?

```// an ephemeral tailnet echo server```

I'm also thinking of netcat and that cats have tails but I puns are failing me

File: data/blog/libtailscale.mdx (hunk 5 above)
Comment (shayne) [reply data/blog/libtailscale.mdx:165]: `//export` is odd but correct. It's specific to cgo and C references to Go.

File: data/blog/libtailscale.mdx [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+operated on through intermediary functions.
+Userspace memory is accessible to the kernel, but because kernels do
+not have a lot of insight into how userspace manages its memory,
+passing a pointer is always done under very carefully specified conditions.
+
+The typical way to refer to a kernel object from userspace is to be
+given a handle.
+The most familiar handle to a Unix user is a
+[file descriptor](https://en.wikipedia.org/wiki/File_descriptor).
+These are process-unique integers managed that refer to an entry table
```
Comment (WalterHub) [inline data/blog/libtailscale.mdx:151]:
```suggestion
These are process-unique integers that refer to an entry table
```

File: data/blog/libtailscale.mdx (hunk 5 above)
Comment (shayne) [reply data/blog/libtailscale.mdx:165]: image

File: data/blog/libtailscale.mdx (hunk 10 above)
Comment (shayne) [reply data/blog/libtailscale.mdx:64]: `//export` while odd but correct. It's specific to cgo and C references to Go.

File: data/blog/libtailscale.mdx (hunk 5 above)
Comment (shayne) [reply data/blog/libtailscale.mdx:165]: It's passable 🤷

File: data/blog/libtailscale.mdx (hunk 10 above)
Comment (WalterHub) [reply data/blog/libtailscale.mdx:64]: I was wondering about "an" vs. "a".

File: data/blog/libtailscale.mdx [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+Where things get interesting is dealing with managed heap memory.
+Every programming language providing some form of memory safety or
+management imposes (often invisible) rules on allocated heap memory.
+The most dramatic are moving garbage collectors, which can pause some
+fraction of a program at any moment, and move a piece of memory,
+rewriting all pointers into that memory.
+This is extremely common as moving GCs are very CPU-efficient for
+programs that allocate a large amount of short-lived heap memory,
+because they enable very efficient
+[generational memory management](<https://en.wikipedia.org/wiki/Tracing_garbage_collection#Generational_GC_(ephemeral_GC)>).
```
Comment (WalterHub) [inline data/blog/libtailscale.mdx:91]:
```suggestion
[generational memory management](https://en.wikipedia.org/wiki/Tracing_garbage_collection#Generational_GC_(ephemeral_GC)).
```

File: data/blog/libtailscale.mdx (hunk 11 above)
Comment (crawshaw) [reply data/blog/libtailscale.mdx:63]: switched to ordinary quotes (I wrote this in Apple Notes and copied it in)

File: data/blog/libtailscale.mdx (hunk 13 above)
Comment (crawshaw) [reply data/blog/libtailscale.mdx:91]: The `<>` was inserted by prettier, which is step in the CI for this website. (I'd never seen it before, apparently it's some markdown trip to deal with the `()` in the link?)

**Comment by crawshaw** [comment]
PTAL

File: data/blog/libtailscale.mdx (hunk 13 above)
Comment (WalterHub) [reply data/blog/libtailscale.mdx:85]: Interesting!

**Review by WalterHub — APPROVED** [review]

**Review by thisisparker — APPROVED** [review]

**Pull request merged**